### ISSUE TITLE
Would you accept a feature request/PR for adding a deployment pool?

### ISSUE THREAD

**Issue by tristanbarcelon** [Dejulia489/AzurePipelinesPS#10]
I'm thinking of creating a function on my end that would use Add-APDeploymentGroup that behaves as easily as the Azure Devops UI for deployment groups. Specifically, when you create a new Deployment Group, it would:

- automatically create a deployment pool with name: team project + '-' + deployment group name if it does not exist, with poolType = deployment and autoSize = true
- take the pool id of that deployment pool and use it when creating the new deployment group

To accomplish this, I would need to utilize the following api in conjunction with Add-APDeploymentGroup function:
- [Get Agent Pool](https://docs.microsoft.com/en-us/rest/api/azure/devops/distributedtask/pools/get%20agent%20pools?view=azure-devops-rest-5.1)
- [Add Agent Pool](https://docs.microsoft.com/en-us/rest/api/azure/devops/distributedtask/pools/add?view=azure-devops-rest-5.1)

Would you be open to accepting a PR implementing only a subset of the parameters? For example, there's not enough information or example on how to utilize the properties parameter when getting agent pool, nor is there sufficient information on what properties are required in the request body when adding an agent pool. Under those circumstances, I would only implement them minimally to match the back end behavior when adding a deployment group from UI.

Thank you in advance.

**Comment by Dejulia489** [comment]
I have been contemplating something like a AzurePipelinesPSToolkit module that would be home to functions like the one you described. I have a lot of scripts/functions that utilize the api 'base' functions. If that is something you are interested in I will spin up the project and we can add it there.

**Comment by tristanbarcelon** [comment]
Thanks @Dejulia489 . Did you intend on creating another module containing functions for these 'base' functions or would you just store them in the Private folder of this module? Since I have not written Pester tests before, would you be able to guide me? Do these tests talk to an actual Azure Devops instance at build time to confirm passing tests? On the other hand, I just filed an issue in the github repo for microsoft docs asking for clarification on the add pool api (https://github.com/MicrosoftDocs/vsts-rest-api-specs/issues/280).

**Comment by Dejulia489** [comment]
By base functions I was referring to the AzurePipelinesPS module. The majority of these functions perform a single call to the api and cover the 'base' functionality of the Azure DevOps api.

My thought is that the toolkit module would utilize the AzurePipelinesPS functions to perform multiple actions, like a toolkit script would. I'm not 100% sure I want to keep the 'base' functions separate from the 'toolkit' functions, but it just seems cleaner. 

A few other questions...
1. What are your thoughts on separating the toolkit and base functions?
2. Where did you find AzurePipelinesPS?
3. Can you fork this project and commit your changes to a feature branch so I can get a feel of what you would like to implement?

**Comment by tristanbarcelon** [comment]
- I think having the base functions stored in the same repo as modules works well for the needs of the toolkit so I probably wouldn't suggest it for now. I've used Powershell modules that take dependencies on other modules and that works but it leads to a tangled web. Prime example is AzureRM. It gets harder to uninstall multiple versions of a depended module over time due to references so I end up uninstalling 'AzureRM' and 'AzureRM.*' and simply re-installing AzureRM which would also install all the AzureRM.* modules.

- I found AzurePipelinesPS via powershellgallery as I frequently check here for functions I need before I go off and write one on my own.

- I can fork your project and show you what I intended on implementing. I'm also looking at these libraries (https://docs.microsoft.com/en-us/azure/devops/integrate/concepts/dotnet-client-libraries?view=azure-devops) for insight into the rest api parameters. Hopefully, they used nullable types in some of them.
